spinalcordtoolbox/deepseg/core.py

```diff
@@ -0,0 +1,109 @@
+
+# Default values if not asked during CLI call and if not present in json metadata.
+DEFAULT_THRESHOLD = 0.9
+
+
+class ParamDeepseg:
+    """
+    Parameters for deepseg module.
+    """
+    def __init__(self):
+        self.threshold = None
+        self.keep_largest_object = True
+        self.output_suffix = '_seg'
+        self.remove_temp_files = 1
+        self.verbose = 1
+        # TODO: add threshold, keep_big_obj
+
+
+class PostProcessing:
+    """
+    Deals with post-processing of the segmentation.
+    """
+    def __init__(self, param, metadata):
+        """
+        :param param:
+        :param metadata:
+        """
+        self.param = param
+        self.metadata = metadata
+
+    def threshold(self, nii_seg):
+        """
+        Threshold the prediction. For no prediction, set 'threshold' to 0.
+        """
+        if self.param.threshold:
+            thr = self.param.threshold
+        else:
+            if 'threshold' in self.metadata:
+                thr = self.param.threshold
+            else:
+                logger.warning("'threshold' is not defined in the model json file. Using threshold of: {}".format(
+                    DEFAULT_THRESHOLD))
+                thr = DEFAULT_THRESHOLD
+        if thr:
+            nii_seg = imed.postprocessing.threshold_predictions_nib(nii_seg, thr)
+        return nii_seg
+
+    def keep_largest_object(self, nii_seg):
+        """
+        Only keep largest object
+        """
+        if self.param.keep_largest_object:
+            if self.threshold:
+                # Fetch axis corresponding to superior-inferior direction
+                # TODO: move that code in image
+                affine = nii_seg.get_header().get_best_affine()
+                code = nib.orientations.aff2axcodes(affine)
+                if 'I' in code:
+                    axis_infsup = code.index('I')
+                elif 'S' in code:
+                    axis_infsup = code.index('S')
+                else:
+                    raise ValueError(
+                        "Neither I nor S is present in code: {}, for affine matrix: {}".format(code, affine))
+                nii_seg = imed.postprocessing.keep_largest_object_per_slice_nib(nii_seg, axis=axis_infsup)
+            else:
+                logger.warning("Algorithm 'keep largest object' can only be run on binary segmentation.")
+            return nii_seg
+
+
+def segment_nifti(fname_image, folder_model, param):
+    """
+    Segment a nifti file.
+
+    :param fname_image: str: Filename of the image to segment.
+    :param folder_model: str: Folder that encloses the deep learning model.
+    :param param: class ParamDeepseg: Parameter class ParamDeepseg()
+    :return: fname_out: str: Output filename.
+    """
+
+    nii_seg = imed.utils.segment_volume(folder_model, fname_image)
+
+    # Postprocessing
+    metadata = sct.deepseg.models.get_metadata(folder_model)
+    postproc = PostProcessing(param, metadata)
```

Comment:
@kousu is that an overkill to create a Class for this or should we simply define each postprocessing function with nii, param and metadata as input parameter?

Reply:
I think a class doesn't offer anything here. A class is useful if you need to keep some state around for a while, but here we want to use.

I think what I would do is write it as

```
def postprocess(nii, param, metadata):
   param = {**DEFAULTS, **metadata, **{k: v for k,v in param.__dict__.items() if v is not None}}
   
   if param['keep_largest_object']:
      nii_seg = keep_largest_object(nii_sig)
   ...
```

Reply:
the "merge" approach is brilliant. Currently trying to implement it.

Reply:
implemented in bea7a3cb50862cfebe64d36dc71647cf4a7ae1dc